lib/SILOptimizer/Mandatory/DiagnoseVarUsage.cpp

```diff
@@ -0,0 +1,1593 @@
+        return true;
+      } 
+    }
+    
+    if (auto Callee = I->getAllOperands()[0].get()) {
+      if (Callee == Value) {
+        // This variable is a closure and is being called.
+        return true;
+      } else if (isa<SILFunctionArgument>(Callee)) {
+        // This is calling a closure passed in as a function argument
+        // A read can not be guaranteed in runtime, but can be assumed
+        return true;
+      }
+    }
+    
+    // If this function is declared externally, and we can
+    // inspect argument conventions.
+    return functionArgIsRead(I, Op);
+  }
+  
+  /// Finds the corresponding SILArgument for a parameter in a basic block
+  static SILArgument *getBasicBlockArgInst(BranchInst *BI, SILValue Param) {
+    
+    // Get index of instruction in operand list.
+    unsigned argIndex = 0;
+    if (BI->getArgs().size() < 1) {
+      return nullptr;
+    }
+    
+    // Find the SILValue in list of Operands
+    if (auto Value = BI->getArgs()[argIndex]) {
+      while (Value != Param) {
+        ++argIndex;
+        if (argIndex == BI->getArgs().size()) {
+          return nullptr;
+        }
+        Value = BI->getArgs()[argIndex];
+      }
+    } else {
+      return nullptr;
+    }
+    
+    if (auto *bb = BI->getDestBB()) {
+      if (SILArgument *Arg = bb->getArgument(argIndex)) {
+        return Arg;
+      }
+    }
+    return nullptr;
+  }
+
+  /// Determines if a store instruction terminates the traversal as a read of
+  /// the entry instruction.
+  bool storeInstIsRead(SILInstruction *I, SILValue Value) {
+    auto SI = dyn_cast<StoreInst>(I);
+    if (SI->getSrc() != Value) {
+      return false;
+    }
+    
+    auto DI = SI->getDest().getDefiningInstruction();
+    
+    // FIXME: Assuming this is a store into inout
+    if (!DI) {
+      return true;
+    }
+    
+    // The destination of this store is a var declaration.
+    // Because we are tracking the use of a value, which may be referenced
+    // by multiple var declarations, check in the AST if the assignment
+    // specifically uses the current var declaration.
+    if (auto *DestVD = getVarDecl(DI)) {
+      auto VD = getVarDecl(EntryInst);
+      return isAssignedToVarDecl(VD, DestVD);
+    }
+    
+    return true;
+  }
+  
+  /// Checks if a SILInstruction belongs to any block accessible by the entry block
+  bool inAccessibleBlock(SILInstruction *I) {
+    SmallVector<SILBasicBlock *, 32> BlockList;
+    llvm::SetVector<SILBasicBlock *> SeenBlocks;
+    BlockList.push_back(EntryInst->getParent());
+    SeenBlocks.insert(EntryInst->getParent());
+    while (!BlockList.empty()) {
+      auto *BB = BlockList.pop_back_val();
+      
+      // Check if the instruction is in this block
+      if (I->getParent() == BB)
+        return true;
+      
+      // Search for more branches
+      auto i = BB->begin(), e = BB->end();
+      while (i != e) {
+        if (auto BI = dyn_cast<BranchInst>(&*i)) {
```

Comment:
No need to handle all terminator instructions. You can just do
```
for (SILBasicBlock *succBB : bb.getSuccessorBlocks())
```
And even if you need the terminator instruction for some reason, you don't have to search for it in the whole instruction list of the block. Just use `SILBasicBlock::getTerminator()`